libphobos/libdruntime/gcc/deh.d

```diff
@@ -143,15 +145,56 @@ __gdc_exception_cleanup(_Unwind_Reason_Code code, _Unwind_Exception *exc)
 }
 
 
+// Each thread in a D program has access to a globalExceptions object.
+
+private struct globalExceptions
+{
+  d_exception_header *thrownExceptions;
+}
+
+globalExceptions __globalExceptions;
+
+// Called before starting a catch.  Returns the exception object.
+
+extern(C) void *
+__gdc_begin_catch(void *exc_ptr)
+{
+  _Unwind_Exception *exceptionObject = cast(_Unwind_Exception *) exc_ptr;
+  d_exception_header *header = get_exception_header_from_ue(exceptionObject);
+  globalExceptions *globals = &__globalExceptions;
+
+  // Something went wrong when stacking up chained headers...
+  if (header != globals.thrownExceptions)
+    __gdc_terminate();
+
+  void *objectp = cast(void *) header.object;
+
+  // Handling for this exception is complete.
+  globals.thrownExceptions = header.nextException;
+  _Unwind_DeleteException (&header.unwindHeader);
+
+  return objectp;
+}
+
 // Perform a throw, D style. Throw will unwind through this call,
 // so there better not be any handlers or exception thrown here.
 
 extern(C) void
-_d_throw(Object object)
+_d_throw(Object o)
 {
-  // FIXME: OOM errors will throw recursively.
```

Comment:
I don't think that's realistic.  A malloc'd exception_header holding the only reference to our thrown Exception == Unexpected garbage collection.  Recursive throws can be instead caught with a function guard, such as a thread-local integer that is incremented/decremented.